### ISSUE TITLE
Assert failure: SyncTableEntry::GetSyncTableEntry()[sbIndex].m_Object == obj

### ISSUE THREAD

**Issue by AntonLapounov** [dotnet/runtime#74741]
The failure has been observed on [x64](https://github.com/dotnet/runtime/pull/72082#issuecomment-1184680141) and [ARM64](https://github.com/dotnet/runtime/pull/72082#issuecomment-1227839169) while using server GC:

```
Assert failure: SyncTableEntry::GetSyncTableEntry()[sbIndex].m_Object == obj
```
The sync table entry with index `sbIndex` is in the free list pointed by `m_FreeSyncTableList` near its beginning.

Call stack examples:
```
04 coreclr!DbgAssertDialog
05 coreclr!ObjHeader::Validate
06 coreclr!Object::ValidateInner
07 coreclr!Object::Validate
08 coreclr!SVR::GCHeap::IsPromoted
09 coreclr!CheckPromoted
0a coreclr!ScanConsecutiveHandlesWithoutUserData
0b coreclr!BlockScanBlocksWithoutUserData
0c coreclr!SegmentScanByTypeMap
0d coreclr!TableScanHandles
0e coreclr!HndScanHandlesForGC
0f coreclr!Ref_CheckReachable
10 coreclr!GCScan::GcWeakPtrScan
11 coreclr!SVR::gc_heap::mark_phase
```
```
04 coreclr!DbgAssertDialog
05 coreclr!ObjHeader::Validate
06 coreclr!Object::ValidateInner
07 coreclr!Object::Validate
08 coreclr!SVR::GCHeap::IsPromoted
09 coreclr!ClearDependentHandle
0a coreclr!ScanConsecutiveHandlesWithUserData
0b coreclr!BlockScanBlocksWithUserData
0c coreclr!SegmentScanByTypeChain
0d coreclr!TableScanHandles
0e coreclr!HndScanHandlesForGC
0f coreclr!Ref_ScanDependentHandlesForClearing
10 coreclr!GCScan::GcWeakPtrScan
11 coreclr!SVR::gc_heap::mark_phase
```

**Comment by AntonLapounov** [comment]
@PeterSolMS I suspect this may be caused by #41600, which made it possible for one thread to start scanning for deleted entries in the `SyncBlockCache` while another thread is still performing `GCScan::GcWeakPtrScan`.

**Comment by PeterSolMS** [comment]
@AntonLapounov I think you are right - I'm thinking about the proper way to fix this...

**Comment by PeterSolMS** [comment]
This should be fixed by PR #74810. Thanks @AntonLapounov for the thorough diagnosis!

**Comment by AntonLapounov** [comment]
Thank you for fixing this!
